### PR TITLE
VEN-1459: Support for fee supporting tokens in shortfall

### PR DESCRIPTION
**Pull request by defcon022**
## Description

Resolves #

## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: contracts/Shortfall/Shortfall.sol [hunk 1]
```diff
@@ -178,19 +179,23 @@ contract Shortfall is Ownable2StepUpgradeable, AccessControlledV8, ReentrancyGua
...
-                        MAX_BPS);
-                    erc20.safeTransfer(auction.highestBidder, previousBidAmount);
+                    erc20.safeTransfer(auction.highestBidder, auction.bidAmount[auction.markets[i]]);
                 }
 
                 uint256 currentBidAmount = ((auction.marketDebt[auction.markets[i]] * bidBps) / MAX_BPS);
+                uint256 balanceBefore = erc20.balanceOf(address(this));
                 erc20.safeTransferFrom(msg.sender, address(this), currentBidAmount);
+                uint256 balanceAfter = erc20.balanceOf(address(this));
+                auction.bidAmount[auction.markets[i]] = balanceAfter - balanceBefore;
```
[the remark this answers is not in the document]
Comment (defcon022) [reply contracts/Shortfall/Shortfall.sol:189]: Yes, it will always be true.

**Review by chechu — COMMENTED** [review]
LGTM

File: contracts/Shortfall/Shortfall.sol [hunk 2]
```diff
@@ -178,19 +179,23 @@ contract Shortfall is Ownable2StepUpgradeable, AccessControlledV8, ReentrancyGua
...
             } else {
                 if (auction.highestBidder != address(0)) {
-                    erc20.safeTransfer(auction.highestBidder, auction.marketDebt[auction.markets[i]]);
+                    erc20.safeTransfer(auction.highestBidder, auction.bidAmount[auction.markets[i]]);
                 }
 
+                uint256 balanceBefore = erc20.balanceOf(address(this));
                 erc20.safeTransferFrom(msg.sender, address(this), auction.marketDebt[auction.markets[i]]);
+                uint256 balanceAfter = erc20.balanceOf(address(this));
+                auction.bidAmount[auction.markets[i]] = balanceAfter - balanceBefore;
```
[the remark this answers is not in the document]
Comment (kkirka) [reply contracts/Shortfall/Shortfall.sol:198]: I'd say the first 3 lines are `_transferIn`, isn't it?
```
/**
 * @notice Transfers tokens from a user to this contract; returns the actually transferred value (accounting for potential fees)
 * @param erc20 Token
 * @param user User
 * @param amount Amount to transfer
 * @return Actually transferred amount
 */
function _transferIn(IERC20Upgradeable erc20, address user, uint256 amount) returns (uint256) {
    uint256 balanceBefore = erc20.balanceOf(address(this));
    erc20.safeTransferFrom(msg.sender, address(this), auction.marketDebt[auction.markets[i]]);
    uint256 balanceAfter = erc20.balanceOf(address(this));
    return balanceAfter - balanceBefore;
}
```

File: tests/hardhat/Shortfall.ts [hunk 3]
```diff
@@ -614,3 +629,108 @@ describe("Shortfall: Tests", async function () {
...
+    const afterBidderBalance = await mockFloki.balanceOf(bidder1.address);
+    const afterBidder2Balance = await mockFloki.balanceOf(bidder2.address);
+
+    expect(previousFlokiBalance).equal(convertToUnit("99", 18));
+    expect(previousBidderBalance).equal(convertToUnit("98", 18));
+    expect(previousBidder2Balance).equal(convertToUnit("198", 18));
+
+    expect(afterFlokiBalance).equal(convertToUnit("99", 18));
+    expect(afterBidderBalance).equal("196010000000000000000");
+    expect(afterBidder2Balance).equal(convertToUnit("98", 18));
```
Comment (kkirka) [inline tests/hardhat/Shortfall.ts:696]: There's some cool stuff in hardhat-chai-matchers:

```
const tx = await shortfall.connect(bidder2).placeBid(poolAddress, "120");
await expect(tx).to.changeTokenBalance(mockFloki, bidder1.address, expectedBalanceDelta);
```

**Review by kkirka — DISMISSED** [review]

File: tests/hardhat/Shortfall.ts (hunk 3 above)
Comment (defcon022) [reply tests/hardhat/Shortfall.ts]: Thanks @kkirka , I have updated the tests: 492f60496d8d9c38be3e377cb4816a2e6390227b

**Review by chechu — APPROVED** [review]

**Review by GitGuru7 — APPROVED** [review]

**Pull request merged**